apps/web/src/utils/kv/kv.ts

```diff
@@ -0,0 +1,175 @@
+import Redis, { type Redis as RedisType } from 'ioredis';
+import { isDevelopment } from 'apps/web/src/constants';
+import { logger } from '../logger';
+
+type KvConstructorParam =
+  | { url: string; tls?: boolean }
+  | { host: string; port: number; username?: string; password?: string; tls?: boolean };
+
+/**
+ * Provides a limited, type-safe interface to Redis operations.
+ * Intentionally restricts access to dangerous commands and raw client operations.
+ */
+export class KVManager {
+  private client: RedisType | null = null;
+
+  private readonly connectionArg: KvConstructorParam;
+
+  private readonly connectionTls: boolean;
+
+  constructor(connectionParam: KvConstructorParam) {
+    if (!connectionParam || (!('url' in connectionParam) && !('host' in connectionParam))) {
+      throw new Error('No URL or options provided to KVManager');
+    }
+    this.connectionArg = connectionParam;
+    this.connectionTls = connectionParam.tls ?? false;
+  }
+
+  private async getClient(): Promise<RedisType> {
+    if (!this.client) {
+      console.log(
+        'creating new redis client: ',
+        'url' in this.connectionArg ? this.connectionArg.url : this.connectionArg.host,
+      );
+      if (!this.connectionArg) {
+        throw new Error('No URL or options provided to KVManager');
+      }
+
+      if ('url' in this.connectionArg) {
+        this.client = new Redis(this.connectionArg.url, this.connectionTls ? { tls: {} } : {});
+      } else {
+        this.client = new Redis({
+          ...this.connectionArg,
+          tls: this.connectionTls ? {} : undefined,
+        });
+      }
+
+      console.log('redis client created', this.client);
+    }
+
+    try {
+      console.log(`pinging ${this.connectionArg}`);
+      const pingRes = await this.client.ping();
+      console.log('ping response', pingRes);
+    } catch (err) {
+      if (!isDevelopment) {
+        logger.error('KV connection failed', err);
+      }
+      console.error(err);
+      throw new Error(`Failed to connect to KV: ${err}`);
+    }
+
+    return this.client;
+  }
+
+  async ping() {
+    if (this.client) {
+      try {
+        const pingRes = await this.client.ping();
+        return pingRes;
+      } catch (err) {
+        if (!isDevelopment) {
+          logger.error('Failed to scan keys', err);
+        }
+        console.error(err);
+        throw new Error(`Failed to ping: ${err}`);
+      }
+    }
+  }
+
+  async close() {
+    if (this.client) {
+      await this.client.quit();
+      this.client = null;
+    }
+  }
+
+  async scan(cursor: number | string = '0', batchSize: number | string = 10) {
+    try {
+      const client = await this.getClient();
+      const [newCursor, elements] = batchSize
+        ? await client.scan(cursor, 'COUNT', batchSize)
+        : await client.scan(cursor);
```

Comment:
I think you could drop the `await` repetition

Reply:
on this one i either need to put await in both of those spots or have an intermediate result that i await later

typescript error if no awaits:
![image](https://github.com/user-attachments/assets/41de79ba-44b6-471d-a56d-fd5ae5bbb64c)

alternative option
![image](https://github.com/user-attachments/assets/52596307-2eb0-4c97-8797-27a40d1dca19)

let me know if you have strong preference for the alternative and i'll update